### PR TITLE
Update ESLint configuration and packages

### PR DESCRIPTION
**Pull request by Gudahtt**
The ESLint configuration and all related dependencies have been updated. Most changes are related to the JSDoc lint rules; many JSDoc blocks were added or adjusted. All other lint changes were applied automatically using `yarn lint:fix`.

### CODE REVIEW

File: src/changelog.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ const changelogDescription = `All notable changes to this project will be docume
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).`;
 
-interface ReleaseMetadata {
+type ReleaseMetadata = {
```
Comment (vthomas13) [inline src/changelog.ts:16]: I've seen this change made in a few commits and I'm curious about the motivation behind the shift from interface to type. Is this something we are slowly changing across the project, or is there a specific need that we are addressing?

Comment (Gudahtt) [reply src/changelog.ts:16]: See here for more context: https://github.com/MetaMask/eslint-config/pull/216 (and [this Slack thread](https://consensys.slack.com/archives/G1L7H42BT/p1638221772259900) for even more context)

Primarily it was for consistency. No specific need being addressed. Erik referenced some difficulties encountered working with interfaces and our `Json` type, but I can't recall the details on that.

Comment (vthomas13) [reply src/changelog.ts:16]: This is great, thanks for the context @Gudahtt!

Comment (Mrtenz) [reply src/changelog.ts:16]: > Erik referenced some difficulties encountered working with interfaces and our `Json` type, but I can't recall the details on that.

It has to do with interfaces not having an index signature, IIRC.

File: src/changelog.ts [hunk 2]
```diff
@@ -56,6 +63,16 @@ function stringifyCategory(category: ChangeCategory, changes: string[]) {
...
 }
 
+/**
+ * Stringify a changelog release section.
+ *
+ * @param version - The release version.
+ * @param categories - The categories of changes included in this release.
+ * @param options - Additional release options.
+ * @param options.date - The date of the release.
+ * @param options.status - The status of the release (e.g. "DEPRECATED").
```
Comment (Mrtenz) [inline src/changelog.ts:73]: Nit:
```suggestion
 * @param options.status - The status of the release (e.g., "DEPRECATED").
```

File: src/changelog.ts [hunk 3]
```diff
@@ -93,18 +117,49 @@ function stringifyReleases(
...
 function withTrailingSlash(url: string) {
   return url.endsWith('/') ? url : `${url}/`;
 }
 
+/**
+ * Get the GitHub URL for comparing two git commits.
+ *
+ * @param repoUrl - The URL for the GitHub repository.
+ * @param firstRef - A reference (e.g. commit hash, tag, etc.) to the first commit to compare.
+ * @param secondRef - A reference (e.g. commit hash, tag, etc.) to the second commit to compare.
```
Comment (Mrtenz) [inline src/changelog.ts:136]: Nit:
```suggestion
 * @param firstRef - A reference (e.g., commit hash, tag, etc.) to the first commit to compare.
 * @param secondRef - A reference (e.g., commit hash, tag, etc.) to the second commit to compare.
```

File: src/changelog.ts [hunk 4]
```diff
@@ -209,15 +264,15 @@ export default class Changelog {
...
    *
-   * @param options
+   * @param options - Release options.
    * @param options.addToStart - Determines whether the change is added to the
    * top or bottom of the list of changes in this category. This defaults to
    * `true` because changes should be in reverse-chronological order. This
    * should be set to `false` when parsing a changelog top-to-bottom.
    * @param options.date - An ISO-8601 formatted date, representing the release
    * date.
    * @param options.status - The status of the release (e.g. 'WITHDRAWN',
```
Comment (Mrtenz) [inline src/changelog.ts:274]: Nit:
```suggestion
   * @param options.status - The status of the release (e.g., 'WITHDRAWN',
```

File: src/cli.ts [hunk 5]
```diff
@@ -40,6 +40,12 @@ formatting is correct. Verification of the contents is left for manual review.`;
 // eslint-disable-next-line node/no-process-env
 const npmPackageVersion = process.env.npm_package_version;
 
+/**
+ * Determine whether the given URL is valid.
+ *
+ * @param proposedUrl - The URL to validate.
+ * @returns True if the URl is valid, false otherwise.
```
Comment (Mrtenz) [inline src/cli.ts:47]: Nit:
```suggestion
 * @returns True if the URL is valid, false otherwise.
```

File: src/cli.ts [hunk 6]
```diff
@@ -152,6 +201,12 @@ look for changes since the last release (defaults to the entire repository at \
 the current working directory), and where the changelog path is resolved from \
 (defaults to the current working directory).`;
 
+/**
+ * Configure options that are common to all commands.
+ *
+ * @param _yargs - The yargs instance to configure.
+ * @returns A yarn instance configured with all common commands.
```
Comment (Mrtenz) [inline src/cli.ts:208]:
```suggestion
 * @returns A Yargs instance configured with all common commands.
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]
Looks good overall. Just some nits.

File: src/cli.ts (hunk 6 above)
Comment (Gudahtt) [reply src/cli.ts:208]: Haha, oops! Good find.

File: src/changelog.ts (hunk 3 above)
Comment (Gudahtt) [reply src/changelog.ts:136]: TIL that this is the proper way to format it in US English. In Canada we don't use the comma.

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**